build-tools/packages/build-tools/src/type-test-generator/main.ts

```diff
@@ -0,0 +1,177 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
+import path from "node:path";
+import { Project, SourceFile } from "ts-morph";
+import { BrokenCompatTypes } from "../common/fluidRepo";
+import { PackageJson } from "../common/npmPackage";
+import { buildTestCase, TestCaseTypeData } from "../typeValidator/testGeneration";
+import { getFullTypeName, getNodeTypeData, TypeData } from "../typeValidator/typeData";
+
+// Do not check that file exists before opening:
+// Doing so is a time of use vs time of check issue so opening the file could fail anyway.
+// Do not catch error from opening file since the default behavior is fine (exits process with error showing useful message)
+const packageFile = readFileSync("package.json");
```

Comment:
For now this is intended to be run as a binary, not a library, so its not exported in the package level exports. Since it uses all synchronous APIs, if something like flub wanted to run it on every project it shouldn't do it as a library anyway.

If we decide to have a tool that invokes this code as a library, then this would want to change in a few ways (become async, not care what the working directory is, be exported from the package etc). For now I'd like to leave it as is since I don't see a need to use this as a library.

I'm not opposed to making it usable as a library later, I just don't see the benefit, and we would need to change more than just wrapping this in a function to do it properly.

Reply:
It needs to be callable as a library. As it is, it's essentially the same as the old bin `fluid-type-validator`, which we deprecated some months ago. I really don't want to recreate that situation. One practical drawback to this approach is that it will require additional changes to teach fluid-build how to parse the new bin.

@fluidframework/build-tools should not contain scripts; it should only contain APIs that we call from elsewhere. The eventual goal is to move everything we use out of @fluidframework/build-tools and remove that package completely. With that in mind, ideally the new code would be in build-cli/src/lib, but putting it there would require adding new exports from @fluidframework/build-tools, so I'm OK leaving the new code as it is.

Reply:
I still don't understand the reason it "It needs to be callable as a library". I can see the argument that something might be more efficient calling it as a library to avoid spawning an extra node process, but I don't see the "need". Also currently as far as I can tell, both the normal build and fluid-build spawn new processes for these kinds of operations (ex: build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts uses execAsync which uses child_process.exec), so I don't think we would even get a performance win in the build by making it usable as a library (and I don't think we need/want it run by anything other than the build)

Regardless, I'd like to consider support for use as a library as a feature that can be implemented later once the code is better factored. This PR is indented as a MVP, and it does work as it.

Reply:
> I still don't understand the reason it "It needs to be callable as a library".

Because we've spent the last 8 months consolidating all of our disparate scripts into one common place -- and this is going the opposite direction. It seems backwards to me to add new code that breaks the "rules" we've been trying to follow while refactoring/updating the build-tools. It's not about perf, or features, or anything like that. It's about consistency in our tooling. In a month this and `fluid-build` will be the only bin entries left in build-tools. In 6 months someone will ask, "Why is this script different? Does it have to be? Should it be?" In a year someone will wrap it as a function and call it from flub.

Reply:
Trying to think of a middle ground option... maybe "callable as a library" is imprecise and thus incorrect. My objections are summarized:

- I don't want to add a new bin entry to @fluidframework/build-tools because we are phasing that library out (partly because it shouldn't be in the @fluidframework scope).
- Adding a bin entry to build-cli is doable, but also odd. What's wrong with `flub generate typetests`?
- I value consistency (maybe too much) as a tool to ease maintenance costs, and I think the consistent "access point" for generating typetests is `flub generate typetests`. This is particularly true in this case because the type tests have two distinct phases, the generation and the prep/config, and prep is run using `flub prepare typetests` (well, it will be -- everything is done in `generate` at the moment).
- Despite my "must be callable as a library" statement, I think it's the "access point" I care most about, not the specifics of how that's achieved.

Reply:
> Trying to think of a middle ground option... maybe "callable as a library" is imprecise and thus incorrect. My objections are summarized:
> 
> * I don't want to add a new bin entry to @fluidframework/build-tools because we are phasing that library out (partly because it shouldn't be in the @fluidframework scope).

We have a different thread about this ( https://github.com/microsoft/FluidFramework/pull/14334#discussion_r1120744873 ) awaiting your response.

> * Adding a bin entry to build-cli is doable, but also odd. What's wrong with `flub generate typetests`?

My objection to this is that it increases coupling, makes the code more fluid specific, is complex for people to understand and maintain (no everyone know how flub works so integrating with it has cost) and adds no value. Also that if we do want to do it, I'd like to not do it in this PR (see note linked above).

This tool is usable on non fluid code. It works with npm, pnpm, or other tools. We could publish it as its own package. People working on this code don't have to know or learn anything about our repo or flub. We lose all these benefits if we make the only way to use it via flub, and I see no benefit to making two ways to use it.

Should tsc be made into a flub command? prettier? eslint? These are all stand alone tools which can easily be run on projects for a reason. They are not bundled into a single-entry point with conceptually unrelated tooling which runs on different scopes and needs different configuration. This is the pattern I'm following.

> * I value consistency (maybe too much) as a tool to ease maintenance costs, and I think the consistent "access point" for generating typetests is `flub generate typetests`. This is particularly true in this case because the type tests have two distinct phases, the generation and the prep/config, and prep is run using `flub prepare typetests` (well, it will be -- everything is done in `generate` at the moment).

The place people look an access point for our tooling is scripts in the package.json files. This is an industry standard we heavily use already. If you want to run the build its in there, same for the linter, test running, formatter etc. Why not include the test generator?

Also I don't see how its really relevant for generating type tests as you should just run the build and they are generated (and tested) as part of the build. I don't think there is much value in running independent of the build, but you easily can with my setup, including all of pnpm's filtering options, or local to the package you are working on: to me this is easier to discover and uses more general/transferable skills than using flub for it. I see no value in replicating pnpm's filtering tools in flub: let per package things be handled by pnpm and repo and release group level things be flub's domain.